### PR TITLE
[repo] Contribution guide

### PR DESCRIPTION
**Pull request by chernser**
## Summary
Reworked contribution guide to tell more about process and expectations. 
We should value every ones time and reduce review cycles so we explain how to prepare PR for best result.

---

> [!NOTE]
> **Low Risk**
> Documentation-heavy changes plus a profile/workflow rename; CI behavior should be equivalent if `-Pcoverage` matches the old `-Panalysis` JaCoCo setup.
> 
> **Overview**
> **Reworks `CONTRIBUTING.md`** into a fuller contribution guide: process (when to open issues/proposals, PR size and scope), testing expectations (including negative tests and coverage quality), a PR checklist (`CHANGELOG.md`, `docs/features.md`, local CI), restricted CI workflow changes, and a structured **Technical Verification** section (build, unit/integration tests, optional native binary and benchmarks).
> 
> **Aligns coverage tooling naming:** the root Maven profile `analysis` is renamed to **`coverage`**, and the GitHub workflow is retitled from **Analysis** to **Coverage** with the Sonar step using `-Pcoverage` instead of `-Panalysis` (same JaCoCo/Sonar flow, clearer name for contributors).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 85907efd0191d80fee3d0039ed00080653b4c58f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by joe-clickhouse — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,125 +1,194 @@
...
+
+Feature work and big changes must be discussed through an issue, and the implementation proposal should be approved before code changes begin. The proposal does not need to be perfect, but it should describe:
+
+- what problem is being solved
+- what user-visible behavior will change
+- what compatibility risks exist
+- what tests will prove the behavior
+- whether documentation, `CHANGELOG.md`, or `docs/features.md` needs to change
+
+It is fine to use AI tools to draft an implementation proposal, explore design options, or prepare a first implementation. AI-generated code is welcome when the specification is detailed, the resulting code follows project patterns, and the tests clearly validate the expected behavior. Special attention should be paid to tests because they are the main guardrails for code changes.
```
Comment (mshustov) [inline CONTRIBUTING.md]: >  AI-generated code is welcome when

should we add `AI_POLICY.md` and link it here?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,125 +1,194 @@
-## Getting started
-ClickHouse-JDBC client is an open-source project, and we welcome any contributions from the community. Please share your ideas, contribute to the codebase, and help us maintain up-to-date documentation.
+# Contributing to ClickHouse Java
+
+Contributions are always welcome: issues, pull requests, questions, ideas, tests, documentation, and examples all help the project.
+
+ClickHouse Java is maintained by a team responsible for the stability and quality of released artifacts. Because this repository publishes client libraries and drivers used by applications, every contribution needs enough context and validation for maintainers to review its safety. At minimum, a code change should include a clear description and tests that prove the intended behavior.
```
Comment (mshustov) [inline CONTRIBUTING.md]: > every contribution needs enough context and validation for maintainers to review its safety

I'd suggest rephrase to something > `here's what makes them easy to review`, otherwise, it reads too defensively

Comment (mshustov) [inline CONTRIBUTING.md]: > Because this repository publishes client libraries and drivers used by applications, every contribution needs enough context and validation for maintainers to review its safety.

I'd even strengthen it 
> The libraries are foundational infrastructure for the software built on top of them, thus a regression or breaking change in a client propagates to applications that depends on it, often with limited ability for users to work around the issue

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,125 +1,194 @@
...
+- what tests will prove the behavior
+- whether documentation, `CHANGELOG.md`, or `docs/features.md` needs to change
+
+It is fine to use AI tools to draft an implementation proposal, explore design options, or prepare a first implementation. AI-generated code is welcome when the specification is detailed, the resulting code follows project patterns, and the tests clearly validate the expected behavior. Special attention should be paid to tests because they are the main guardrails for code changes.
+
+Some good ideas need time to design and implement. We are open to discussion and welcome collaboration in issues, even when the final implementation is not ready yet.
+
+### Big Contributions
+
+Large changes are easier to review and merge when they are split into a few smaller pull requests. From our experience, it is often better to prepare the codebase first with focused cleanup, refactoring, or test-only PRs, and then add the feature or behavior change in a later PR.
```
Comment (mshustov) [inline CONTRIBUTING.md]: The principle is right but the guidance doesn't have an easy to follow definition. Let's give contributors a concrete threshold. For example (numbers are taken from this study https://www.propelcode.ai/blog/pr-size-impact-code-review-quality-data-study):
> As a rough guideline, if a pull request is over 300–400 lines of code (excluding docs and tests), consider whether it can be split. Smaller PRs are reviewed and merged faster, and the savings in review time usually outweigh the overhead of splitting

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,125 +1,194 @@
...
+
+Please make pull requests review-ready before requesting review. A PR may still take time to review and prepare for merge, especially when it changes public behavior or touches multiple modules.
+
+Every pull request should include:
+
+- a clear description of what changed and why
+- a link to the related issue, when one exists
+- an explanation of the user-visible behavior and compatibility impact
+- tests for the changed behavior, including integration tests when the change interacts with ClickHouse server behavior
+- the local test commands that were run
```
Comment (mshustov) [inline CONTRIBUTING.md]: not sure we need that level of noize in every PR. I'd just set expectations: `you run test before opening a PR`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,125 +1,194 @@
...
+
+Test code is as important as production code. Keep tests compact, readable, and focused on the scenario being verified. Avoid repeating existing coverage unless the new test adds a distinct scenario, edge case, module, type, format, or failure mode. Reduce duplication in test setup and assertions so reviewers can quickly see what behavior is being tested.
+
+We measure test coverage, but coverage percentage is not the only goal. We value tests that cover meaningful scenarios and edge cases. For example, a test that reads a random large integer is useful, but a stronger test also checks boundary values, rounding behavior, invalid values, and the interaction with nullable or nested types when relevant.
+
+### Restricted Changes
+
+Changes to CI workflows are currently restricted. Please discuss CI workflow changes with maintainers before opening a pull request.
+
+## Pull Request Checklist
```
Comment (mshustov) [inline CONTRIBUTING.md:68]: I'd add a couple of things to set clearer expectations on both sides:

- CI must be runnable on the PR. For external contributors, the initial CI run requires approval from a maintainer (as of today, probably should be revisited). Let’s explicitly mention this and commit to triggering it promptly to avoid contributors being stuck waiting. For internal PRs, the CI suite should actually run against the proposed change before the review process begins. If tests are scoped down, skipped, or disabled, the PR should clearly indicate this.

- Reviews occur when CI is green. Clearly state that the team will review the code once CI passes. Additionally, contributors should either fix PRs with failing CI or move them to the draft. This approach is fair in both directions: contributors have a clear understanding of what means `ready for review`, and reviewers avoid spending cycles on changes that are not ready

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,125 +1,194 @@
-## Getting started
-ClickHouse-JDBC client is an open-source project, and we welcome any contributions from the community. Please share your ideas, contribute to the codebase, and help us maintain up-to-date documentation.
+# Contributing to ClickHouse Java
```
Comment (mshustov) [inline CONTRIBUTING.md]: Most of the guidance provided in the document (Testing, Contribution Process, significant contributions, and the PR checklist) is also relevant to AI coding agents. Considering the significant portion of our contributions made by agents, we should either mirror these rules into `AGENTS.md` or extract them into a shared file that both documents reference.
This can be done in a separate PR.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,125 +1,194 @@
...
+- what user-visible behavior will change
+- what compatibility risks exist
+- what tests will prove the behavior
+- whether documentation, `CHANGELOG.md`, or `docs/features.md` needs to change
+
+It is fine to use AI tools to draft an implementation proposal, explore design options, or prepare a first implementation. AI-generated code is welcome when the specification is detailed, the resulting code follows project patterns, and the tests clearly validate the expected behavior. Special attention should be paid to tests because they are the main guardrails for code changes.
+
+Some good ideas need time to design and implement. We are open to discussion and welcome collaboration in issues, even when the final implementation is not ready yet.
+
+### Big Contributions
```
Comment (mshustov) [inline CONTRIBUTING.md:40]: One logical change per PR: a feature, a bug fix, a refactor, or a doc update.  Mixed-purpose PRs complicate reviews and increase the likelihood of unintended regressions.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,125 +1,194 @@
...
+
+Large changes are easier to review and merge when they are split into a few smaller pull requests. From our experience, it is often better to prepare the codebase first with focused cleanup, refactoring, or test-only PRs, and then add the feature or behavior change in a later PR.
+
+Avoid packing unrelated cleanup, infrastructure changes, test rewrites, and the feature itself into one large pull request. Smaller PRs help maintainers verify intent, reduce review time, and lower the risk of regressions.
```
Comment (mshustov) [inline CONTRIBUTING.md]: For a new feature, implement the smallest change that solves the problem.  Polish, additional configuration, optimization, and extensions can be addressed in the follow-up PRs

File: CONTRIBUTING.md (hunk 4 above)
Comment (mshustov) [inline CONTRIBUTING.md]: - run the project's code review skill against your diff before requesting human review

this and `run tests locally` can be moved to `before you open a PR`

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,125 +1,194 @@
...
+
+Changes to CI workflows are currently restricted. Please discuss CI workflow changes with maintainers before opening a pull request.
+
+## Pull Request Checklist
+
+Please make pull requests review-ready before requesting review. A PR may still take time to review and prepare for merge, especially when it changes public behavior or touches multiple modules.
+
+Every pull request should include:
+
+- a clear description of what changed and why
```
Comment (mshustov) [inline CONTRIBUTING.md]:
```suggestion
- a clear description of what changed, why and why now
```

File: CONTRIBUTING.md (hunk 6 above)
Comment (abonander) [reply CONTRIBUTING.md]: Can the `AGENTS.md` just say "please also read `CONTRIBUTING.md`"?

File: CONTRIBUTING.md (hunk 2 above)
Comment (abonander) [reply CONTRIBUTING.md]: Maybe

> ClilckHouse client libraries are deployed in production environments all over the world. Our users and customers depend on them being thoughtfully designed, well documented, and highly reliable. The high bar for quality we apply to all our software must apply to external contributions as well.

File: CONTRIBUTING.md (hunk 5 above)
Comment (abonander) [reply CONTRIBUTING.md:68]: I would also mention:

* Out-of-date branches and merge conflicts should be resolved promptly (GitHub should send an email notifying when this happens)
* PRs that remain open for extended periods with unresolved conversations, merge conflicts or failing CI may be closed at maintainer discretion
    *  I learned the hard way that a PR author cannot re-open a PR that they did not close themselves. This should mention that if we close a PR for being stale and they want to come back to it, they will need to open a new PR.

File: CONTRIBUTING.md (hunk 4 above)
Comment (abonander) [reply CONTRIBUTING.md]: This kind of falls under the stale-PR point above. If the user is clearly not running tests before submitting, CI will be consistently failing and that's a fair reason for us to close the PR.

File: CONTRIBUTING.md (hunk 1 above)
Comment (chernser) [reply CONTRIBUTING.md]: yes. Will add link.

File: CONTRIBUTING.md (hunk 2 above)
Comment (chernser) [reply CONTRIBUTING.md]: What about: 

> Libraries in this repository are used in many production systems. Users expect certain level of quality and 
> active support (adding new features, fixing issues). It is important to make all changes safe and swiftly.
> There we ask all contributors for tests and clear description of changes done in a PR. This expedites merge and quadruples value of your contribution.

File: CONTRIBUTING.md (hunk 3 above)
Comment (chernser) [reply CONTRIBUTING.md]: > We all know that large PRs (400+ lines of code) are hard to review and it is more likely to miss a bug. Small changes are appreciated and take less time to verify them. If many changes are needed anyway (feature implementation, refactoring), please split them: something can be done as code preparation, cleanup or smaller improvement. It should take less time at sum as this repository is actively maintained. We will ask to split changes of 800+ LOC anyway.

File: CONTRIBUTING.md (hunk 4 above)
Comment (chernser) [reply CONTRIBUTING.md]: @mshustov 

Ok. I will rephrase 
> - a link to the related issue
> - compatibility impact and behavior changes 
> - unit and integration test 

> Make sure all existing and new tests are passing. PRs with failing CI will stay open.

File: CONTRIBUTING.md (hunk 6 above)
Comment (chernser) [reply CONTRIBUTING.md]: @mshustov 
I agree. Will write something. 

@abonander 
This might be a good idea.

**Comment by joe-clickhouse** [comment]
I really like where this is going. The only thing I'm weary of is that the longer it gets, the less likely a human is going to read the entire thing. An agent will no problem, so the length with detailed justifications and directions is great, but it might be worth having a short bulleted TLDR section at the top that provides md links to the relevant sections? e.g:

```markdown
⚡ TLDR
- Proposals: Small fix? Just open a PR. Big feature or API change? Open an [Issue](#issues-discussion-and-proposals) first.
- PR Requirements: Every PR needs a clear description, tests, and a CHANGELOG.md update. See the full [PR Checklist](#pull-request-checklist).
- Testing: Code changes require both positive and negative tests. Keep them focused. More in [Testing Expectations](#testing-expectations).
- Quick Build Command (JDK 17+): `mvn -DskipITs` clean verify (Requires [Toolchains setup](#set-up-the-environment)).
- Run Local Tests: Unit tests run automatically. For Integration Tests, ensure Docker is running. Learn how to [Run Unit and Integration Tests](#run-unit-and-integration-tests).
```

**Comment by chernser** [comment]
@joe-clickhouse 

thank you for the feedback! 

Document is mainly focused for people not agents. I will add suggested section at header and will try to minimize wording to make document shorter and may be bit more structured.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,125 +1,194 @@
...
+#clickhouseVersion=latest
+#clickhouseTimezone=UTC
+#clickhouseImage=clickhouse/clickhouse-server
+#additionalPackages=
+#proxyImage=ghcr.io/shopify/toxiproxy:2.5.0
+```
+
+TBD: document a dockerized development environment for running the standard local test suite.
+
+### 5. Optional Native Binary Check
```
Comment (alex-clickhouse) [inline CONTRIBUTING.md]: Would it make sense to also have instructions for checking test coverage at this point?

Comment (chernser) [reply CONTRIBUTING.md]: yes, this is good idea. Will add.

File: CONTRIBUTING.md (hunk 1 above)
Comment (chernser) [reply CONTRIBUTING.md]: done.

File: CONTRIBUTING.md (hunk 7 above)
Comment (chernser) [reply CONTRIBUTING.md:40]: Added.

File: CONTRIBUTING.md (hunk 8 above)
Comment (chernser) [reply CONTRIBUTING.md]: added

File: CONTRIBUTING.md (hunk 5 above)
Comment (chernser) [reply CONTRIBUTING.md:68]: done.

File: CONTRIBUTING.md (hunk 9 above)
Comment (chernser) [reply CONTRIBUTING.md]: removed

File: CONTRIBUTING.md (hunk 4 above)
Comment (chernser) [reply CONTRIBUTING.md]: reworked.

Comment (chernser) [reply CONTRIBUTING.md]: reworked.

File: CONTRIBUTING.md (hunk 10 above)
Comment (chernser) [reply CONTRIBUTING.md]: done.

**Review by mshustov — APPROVED** [review]
Nit: I'd still consider adding `tldr;` section as proposed by @joe-clickhouse
